cassiopeia/simulator/SpatialSampler.py

```diff
@@ -0,0 +1,315 @@
+            (min_y, max_y + self.cell_size),
+        )
+        cells = {
+            leaf: self.create_cell(location[0], location[1])
+            for leaf, location in locations.items()
+        }
+
+        # Capture cells in each spot.
+        # This dictionary will contain the index of each spot (in the spots
+        # list) as the keys and a dictionary of captured cells (as returned by
+        # `capture_cells_on_spot`) as values.
+        captured_spots = {}
+        for i, spot in enumerate(spots):
+            captured = self.capture_cells(spot, cells)
+            if captured:
+                captured_spots[i] = captured
+
+        # Construct a new network and character matrix
+        sampled_tree = tree.copy()
+        spot_leaves = {}
+        for spot_i, captured in captured_spots.items():
+            spot_uuid = str(uuid.uuid4())
+            spot_leaves[spot_uuid] = spot_i
+            spot_states = self.sample_states(tree, captured)
+
+            # Add new leaf
+            captured_leaves = list(captured.keys())
+            new_time = sum(
+                tree.get_time(leaf) for leaf in captured_leaves
+            ) / len(captured_leaves)
+            sampled_tree.add_leaf(
+                tree.find_lca(*captured_leaves)
+                if len(captured_leaves) > 1
+                else tree.parent(captured_leaves[0]),
+                spot_uuid,
+                states=spot_states,
+                time=new_time,
+            )
+
+        # Remove all other leaves
+        # Construct new character matrix and cell meta.
+        relabel_map = {}
+        character_matrix = pd.DataFrame(
+            columns=tree.character_matrix.columns, dtype=object
+        )
+        meta_columns = [f"{attribute_key}_0", f"{attribute_key}_1"]
+        cell_meta = pd.DataFrame(columns=meta_columns, dtype=float)
+        for leaf in sampled_tree.leaves:
+            if leaf not in spot_leaves:
+                sampled_tree.remove_leaf_and_prune_lineage(leaf)
+            else:
+                # Set attribute in network
+                spot_center = spots[spot_leaves[leaf]].centroid
+                location = (spot_center.x, spot_center.y)
+                sampled_tree.set_attribute(leaf, attribute_key, location)
+
+                # Add rows to character_matrix and cell_meta
+                new_leaf = f"s{len(relabel_map)}"
+                states = sampled_tree.get_character_states(leaf)
+                states_arr = np.empty(len(states), dtype=object)
+                states_arr[:] = states
+                character_matrix.loc[new_leaf] = states_arr
+                cell_meta.loc[new_leaf, meta_columns] = location
+
+                relabel_map[leaf] = new_leaf
+
+        # Relabel spots to be an integer prefixed with "s" (indicating "spot")
+        sampled_tree.relabel_nodes(relabel_map)
+        sampled_tree.character_matrix = character_matrix
+        sampled_tree.cell_meta = cell_meta
```

Comment:
Currently, the way the `register_data_with_tree` is implemented, for each node not in the character matrix, we add the node with all missing states to the character matrix. We don't attempt to read these character states from the tree. If we changed this, we may be able to just run `register_data_with_tree` at the end of the tree generation. This would also allow us to remove the character matrix handling in the above, as when nodes are removed with `remove_and_prune_lineage` they are removed from all character matrix layers, and each new node would have their states read in from the tree. Let me know what y'all think!